### PR TITLE
Update managed vnet check logic

### LINKED ISSUE #4930: BYO VNet Deletion Issues: Route Tables & NSGs
/kind bug

**What steps did you take and what happened:**
Deployed private cluster using a BYO (Bring Your Own) VNet which caused problems with the deletion of route tables and network security groups (NSGs). According to the comments in the code, this shouldn't be created in the first place.
```
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureCluster
metadata:
  name: ${CLUSTER_NAME}
  namespace: default
spec:
  identityRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: AzureClusterIdentity
    name: ${CLUSTER_IDENTITY_NAME}
  location: ${AZURE_LOCATION}
  networkSpec:
    apiServerLB:
      name: ${CLUSTER_NAME}-internal-lb
      type: Internal
      frontendIPs:
      - name: lb-private-ip-frontend
        privateIP: 10.241.1.138
    controlPlaneOutboundLB:
      frontendIPsCount: 1
    nodeOutboundLB:
      frontendIPsCount: 1
    subnets:
    - name: control-plane-subnet
      role: control-plane
      cidrBlocks:
        - 10.241.1.128/26
      securityGroup:
        name: control-plane-nsg
    - name: node-subnet
      role: node
      cidrBlocks:
        - 10.241.2.192/26
      securityGroup:
        name: node-subnet-nsg
      routeTable:
        name: node-subnet-routetable
    vnet:
      name: ${AZURE_VNET_NAME:=${CLUSTER_NAME}-vnet}
      resourceGroup: rg-io-westeurope-iks-1
  resourceGroup: ${AZURE_RESOURCE_GROUP:=${CLUSTER_NAME}}
  subscriptionID: ${AZURE_SUBSCRIPTION_ID}
```

**What did you expect to happen:**
Network security groups and route tables shouldn't be created when you are using BYO VNet. 

**Anything else you would like to add:**
I think that I've found the root cause, when AzureCluster object is initially created, `VnetId` will be empty and CAPZ will mark Vnet as `managed` and create resources: 
```
I0620 12:55:03.652051      37 cluster.go:589] "Debugging vnet" logger="ClusterScope.IsVnetManaged" x-ms-correlation-request-id="15eff891-9f4c-4d95-b163-9393b17090f2" managed=true vnetID="" vnetTags=null
I0620 12:55:03.771623      37 securitygroups.go:97] "Debugging vnet, creating sg" logger="securitygroups.Service.Reconcile" controller="azurecluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AzureCluster" AzureCluster="default/az-hbasic-debug-1" namespace="default" name="az-hbasic-debug-1" reconcileID="954ba368-2751-467b-9095-01882a41f780" x-ms-correlation-request-id="8a3a8634-2f31-42b8-a6f7-27ce2b3a6544" security group="control-plane-nsg"
I0620 12:55:04.239271      37 securitygroups.go:97] "Debugging vnet, creating sg" logger="securitygroups.Service.Reconcile" controller="azurecluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AzureCluster" AzureCluster="default/az-hbasic-debug-1" namespace="default" name="az-hbasic-debug-1" reconcileID="954ba368-2751-467b-9095-01882a41f780" x-ms-correlation-request-id="8a3a8634-2f31-42b8-a6f7-27ce2b3a6544" security group="node-subnet-nsg"
I0620 12:55:04.337309      37 routetables.go:91] "Debugging vnet, creating route table" logger="routetables.Service.Reconcile" controller="azurecluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AzureCluster" AzureCluster="default/az-hbasic-debug-1" namespace="default" name="az-hbasic-debug-1" reconcileID="954ba368-2751-467b-9095-01882a41f780" x-ms-correlation-request-id="8a3a8634-2f31-42b8-a6f7-27ce2b3a6544" route table="node-subnet-routetable"
```
After some time,  `VnetId` will be populated and Vnet will be marked as `not-managed`. 
```
I0620 12:55:20.636150      37 cluster.go:589] "Debugging vnet" logger="ClusterScope.IsVnetManaged" x-ms-correlation-request-id="ea9a69bb-a738-4ed2-9436-2f59f57fd296" managed=false vnetID="/subscriptions/292c8272-9273-43c4-b998-22a124e86546/resourceGroups/rg-io-westeurope-iks-1/providers/Microsoft.Network/virtualNetworks/vnet-io-westeurope-2" vnetTags=null
```
This could also be the root cause for this issue: #4626 

I'm currently testing a fix for this. It seems we should check if the VNet is managed, similar to how we check for managed clusters.
If you agree I would like to create PR for this issue.  

**Environment:**

- cluster-api-provider-azure version: v1.15.0
- Kubernetes version: (use `kubectl version`): tested on v1.30.0 and v1.27.3
- OS (e.g. from `/etc/os-release`): 
```
NAME="Rocky Linux"
VERSION="8.9 (Green Obsidian)"
ID="redhat"
ID_LIKE="rhel centos fedora"
VERSION_ID="8.9"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Rocky Linux 8.9 (Green Obsidian)"
ANSI_COLOR="0;32"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:rocky:rocky:8:GA"
HOME_URL="https://rockylinux.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
SUPPORT_END="2029-05-31"
ROCKY_SUPPORT_PRODUCT="Rocky-Linux-8"
ROCKY_SUPPORT_PRODUCT_VERSION="8.9"
REDHAT_SUPPORT_PRODUCT="Rocky Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="8.9"
```

### PR DESCRIPTION
**Pull request by hrbasic**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
In some cases, using a BYO (Bring Your Own) VNet can cause problems with the deletion of route tables and network security groups due to an issue with checking if the VNet is managed. More information can be found in the related issue.

**Which issue(s) this PR fixes**:
Fixes #4930

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
```release-note
Refactor method for checking if a virtual network is managed
```

### CODE REVIEW

File: azure/scope/cluster.go [hunk 1]
```diff
@@ -582,9 +582,21 @@ func (s *ClusterScope) IsVnetManaged() bool {
...
+
+	vnet := s.VNetSpec().ResourceRef()
+	vnet.SetNamespace(s.ASOOwner().GetNamespace())
+	err := s.Client.Get(ctx, client.ObjectKeyFromObject(vnet), vnet)
+	if err != nil {
+		log.Error(err, "Unable to determine if ClusterScope VNET is managed by capz, assuming unmanaged", "AzureCluster", s.ClusterName())
+		return false
+	}
+
+	isManaged := infrav1.Tags(vnet.Status.Tags).HasOwned(s.ClusterName())
```
[the remark this answers is not in the document]
Comment (willie-yao) [reply azure/scope/cluster.go:597]: That should be fine, thanks for the explanation!

**Review by willie-yao — APPROVED** [review]
/lgtm
/assign @mboersma

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

**Pull request merged**